### PR TITLE
Inlined attachments

### PR DESCRIPTION
**Pull request by chetanmeh**
This PR enables inlining attachment content as part of attachment name for smaller attachments

## Description

Currently [runtime manifest][1] determine if the code related to any runtime is stored as attachment or as string property as part of action document itself. As part of #3450 plan is introduce a new `AttachmentStore` SPI and move code storage to a separate store. In doing that reading an action with code would invoke 2 remote calls. 

To reduce this overhead for smaller codes this PR enables inlining of attachments as part of attachment name itself.

Under this if any   

1. attachment length \>= `maxInlineSize` - Then its saved as proper attachment
2. attachment length < `maxInlineSize` - Then its content would be converted to URL encoded base64 string and would be stored as part of `attachmentName` within the action document. Thus it can read without going to attachment storage

Attachment name for an inlined attachment would be a URI like below

    mem:T3BlbldoaXNrIEZUVw==

Here 

* Scheme - `mem`
* Path - Would be base64 encoded attachment content

### Defaults

This PR sets the default value of `max-inline-size` to 16 KB

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [X] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

[1]: https://github.com/apache/incubator-openwhisk/blob/master/ansible/files/runtimes.json
[2]: https://console.bluemix.net/docs/services/Cloudant/offerings/bluemix.html#request-and-document-size-limits

### CODE REVIEW

File: common/scala/src/main/resources/application.conf [hunk 1]
```diff
@@ -114,6 +114,14 @@ whisk {
...
         activations-ddoc = "whisks.v2.1.0"
         activations-filter-ddoc = "whisks-filters.v2.1.0"
+
+        # Size limit for inlined attachments. Attachments having size less than this would
+        # be inlined with there content encoded in attachmentName
+        max-inline-size = 16 k
+
+        # Chunk sized for converting source of bytes to ByteString as part of attachment
+        # upload flow
+        chunk-size = 8 k
```
Comment (chetanmeh) [inline common/scala/src/main/resources/application.conf]: Value of 8k is choosed based on default used by [StreamConverters.fromInputStream][1] which is used by `DocumentFactory` to create a `Source[ByteString,_]` from `InputStream` currently

[1]: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala#L48

File: common/scala/src/main/resources/application.conf [hunk 2]
```diff
@@ -114,6 +114,14 @@ whisk {
         actions-ddoc = "whisks.v2.1.0"
         activations-ddoc = "whisks.v2.1.0"
         activations-filter-ddoc = "whisks-filters.v2.1.0"
+
+        # Size limit for inlined attachments. Attachments having size less than this would
+        # be inlined with there content encoded in attachmentName
+        max-inline-size = 16 k
```
Comment (rabbah) [inline common/scala/src/main/resources/application.conf]: can this be set per deployment?

Comment (chetanmeh) [reply common/scala/src/main/resources/application.conf]: It should be possible to override the default via env variable `CONFIG_whisk_db_maxInlineSize`. Is that what you meant?

Comment (rabbah) [reply common/scala/src/main/resources/application.conf]: Yes - great.

**Comment by rabbah** [comment]
Thanks for adding the additional tests. 

Needs a PG.

File: common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala [hunk 3]
```diff
@@ -356,13 +356,21 @@ class CouchDbRestStore[DocumentAbstraction <: DocumentSerializer](dbProtocol: St
...
-    val updatedDoc = update(d, attached)
-
     for {
-      i1 <- put(updatedDoc)
-      i2 <- attach(i1, attachmentUri.path.toString(), attached.attachmentType, docStream)
+      (bytes, tailSource) <- inlineAndTail(docStream)
+      uri <- Future.successful(uriOf(bytes, UUID().asString))
+      attached <- {
+        val a = if (isInlined(uri)) {
+          Attached(uri.toString(), contentType, Some(bytes.size), Some(digest(bytes)))
```
[the remark this answers is not in the document]
Comment (chetanmeh) [reply common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: Good point. Updated now

**Comment by csantanapr** [comment]
PG1 2996 ⌛️

**Comment by csantanapr** [comment]
PG1 2996 👍

**Pull request merged**